src/plugins/event_annotation/common/fetch_event_annotations/request_event_annotations.ts

```diff
@@ -0,0 +1,352 @@
+          size: ANNOTATIONS_PER_BUCKET,
+          sortOrder: 'asc',
+          sortField: group.timeField,
+        },
+      },
+    };
+
+    const fieldsTopMetric = (group.allFields || [])?.map((field) => ({
+      type: 'agg_type',
+      value: {
+        enabled: true,
+        type: 'top_metrics',
+        params: {
+          field,
+          size: ANNOTATIONS_PER_BUCKET,
+          sortOrder: 'asc',
+          sortField: group.timeField,
+        },
+      },
+    }));
+
+    const aggregations = [
+      annotationsFilters,
+      dateHistogram,
+      count,
+      timefieldTopMetric,
+      ...fieldsTopMetric,
+    ];
+
+    const aggConfigs = aggs.createAggConfigs(dataView, aggregations?.map((agg) => agg.value) ?? []);
+    return {
+      esaggsParams: { dataView, aggConfigs, timeFields: [group.timeField] },
+      fieldsColIdMap:
+        group.allFields?.reduce<Record<string, string>>(
+          (acc, fieldName, i) => ({
+            ...acc,
+            [fieldName]: `col-${i + 4}-${i + 5}`,
+          }),
+          {}
+        ) || {},
+    };
+  });
+};
+
+function regroupForRequestOptimization(
+  { groups }: FetchEventAnnotationsArgs,
+  input: ExpressionValueSearchContext | null
+) {
+  const outputGroups = groups
+    .map((g) => {
+      return g.annotations.reduce<Record<string, ManualGroup | QueryGroup>>((acc, current) => {
+        if (current.isHidden) {
+          return acc;
+        }
+
+        if (isManualAnnotation(current)) {
+          if (!isInRange(current, input?.timeRange)) {
+            return acc;
+          }
+          const manualSubgroup = acc.manual as ManualGroup | undefined;
+          return {
+            ...acc,
+            manual: {
+              type: 'manual',
+              annotations: [...(manualSubgroup ? manualSubgroup.annotations : []), current],
+            },
+          };
+        } else {
+          const key = `${g.dataView.value.id}-${current.timeField}`;
+          const subGroup = acc[key] as QueryGroup;
+          if (subGroup) {
+            let allFields = [...(subGroup.allFields || []), ...(current.extraFields || [])];
+            if (current.textField) {
+              allFields = [...allFields, current.textField];
+            }
+            return {
+              ...acc,
+              [key]: {
+                ...subGroup,
+                allFields: [...new Set(allFields)],
+                annotations: [...subGroup.annotations, current],
+              },
+            };
+          }
+
+          let allFields = current.extraFields || [];
+          if (current.textField) {
+            allFields = [...allFields, current.textField];
+          }
+
+          return {
+            ...acc,
+            [key]: {
+              type: 'query',
+              dataView: g.dataView,
+              timeField: current.timeField,
+              allFields,
+              annotations: [current],
+            },
+          };
+        }
+      }, {});
```

Comment:
I think it is safe to mutate `acc` within the `reduce` scope to make the code a little bit more succinct and performant.

In general I would avoid the use of spread vs just push in situations like `allFields`, where the number of fields might be higher - i.e. thinking about the merging of all annotations together.

As for the `new Set`, I would move it at the `reduce` block below when merging the `requestGroup.annotations` together with the `currentSubgroup.annotations` - see comment below.

```suggestion
        if (isManualAnnotation(current)) {
          if (!isInRange(current, input?.timeRange)) {
            return acc;
          }
          if (!acc.manual) {
            acc.manual = { type: 'manual', annotations: [] };
          }
          (acc.manual as ManualGroup).annotations.push(current);
          return acc;
        }
        const key = `${g.dataView.value.id}-${current.timeField}`;
        const subGroup = acc[key] as QueryGroup;
        if (subGroup) {
          if (!subGroup.allFields) {
            subGroup.allFields = [];
          }
          subGroup.allFields.push(...(current.extraFields || []));
          if (current.textField) {
            subGroup.allFields.push(current.textField);
          }
          subGroup.annotations.push(current);
          return acc;
        }

        const allFields = current.extraFields || [];
        if (current.textField) {
          allFields.push(current.textField);
        }

        return {
          ...acc,
          [key]: {
            type: 'query',
            dataView: g.dataView,
            timeField: current.timeField,
            allFields,
            annotations: [current],
          },
        };
      }, {});
```